File: v4/rust/iprange-livedb/src/validation/membership.rs
use sha2::{Digest, Sha256};

use crate::contract::{u32_le, ValueKind, PAGE_SIZE};
use crate::error::{Error, Result};
use crate::format::page_type;
use crate::mapping::{ByteRange, ByteSource};
use crate::membership_view;

use super::bitmap::{self, Kind, WordCache};
use super::blob;
use super::context::Context;
use super::membership_table::InsertResult;
use super::tree::{self, CellLayout, Codec};
use super::{ValidationObject, ValidationReason, ValidationSink};

mod record;

use record::{decode_hash, decode_record, HashKey, Record, Storage};

const ID_BASE: usize = 64;
const MAX_ID_RECORD: usize = PAGE_SIZE - 34;
const COMPARE_WORDS: usize = 64;

struct IdCodec;

impl Codec for IdCodec {
    type Key = u32;

    const BRANCH_TYPE: u8 = page_type::MEMBERSHIP_ID_BRANCH;
    const LEAF_TYPE: u8 = page_type::MEMBERSHIP_ID_LEAF;
    const AUX: u32 = 0;
    const BRANCH_LAYOUT: CellLayout = CellLayout::Fixed(8);
    const LEAF_LAYOUT: CellLayout = CellLayout::Variable {
        minimum: ID_BASE,
        maximum: MAX_ID_RECORD,
    };
    const LEAF_INVALID: ValidationReason = ValidationReason::MembershipBitmapInvalid;

    fn branch_key<P: ByteSource>(cell: P) -> Option<Self::Key> {
        Some(u32_le(cell, 0))
    }

    fn branch_child<P: ByteSource>(cell: P) -> Option<u32> {
        Some(u32_le(cell, 4))
    }

    fn leaf_key<P: ByteSource>(cell: P) -> Option<Self::Key> {
        decode_record(cell).map(|record| record.id)
    }
}

struct HashCodec;

impl Codec for HashCodec {
    type Key = HashKey;

    const BRANCH_TYPE: u8 = page_type::MEMBERSHIP_HASH_BRANCH;
    const LEAF_TYPE: u8 = page_type::MEMBERSHIP_HASH_LEAF;
    const AUX: u32 = 0;
    const BRANCH_LAYOUT: CellLayout = CellLayout::Fixed(44);
    const LEAF_LAYOUT: CellLayout = CellLayout::Fixed(40);
    const BRANCH_INVALID: ValidationReason = ValidationReason::MembershipHashInvalid;
    const LEAF_INVALID: ValidationReason = ValidationReason::MembershipHashInvalid;

    fn branch_key<P: ByteSource>(cell: P) -> Option<Self::Key> {
        decode_hash(ByteRange::new(cell, 0, 40)?)
    }

    fn branch_child<P: ByteSource>(cell: P) -> Option<u32> {
        Some(u32_le(cell, 40))
    }

    fn leaf_key<P: ByteSource>(cell: P) -> Option<Self::Key> {
        decode_hash(cell)
    }
}

pub(crate) fn validate<S: ValidationSink>(context: &mut Context<'_, S>) -> Result<()> {
    if context.meta.value_kind != ValueKind::Membership {
        return Ok(());
    }

    let mut feeds = WordCache::new(
        context.meta.feed_used_root,
        context.meta.feed_index_limit,
        Kind::Feed,
    );
    let mut maximum_id = 0u32;
    let id_result = tree::walk::<IdCodec, S, _>(
        context,
        context.meta.membership_id_root,
        ValidationObject::MembershipDictionary,
        |context, page, cell| validate_record(context, page, cell, &mut feeds, &mut maximum_id),
    )?;
    if id_result.records != context.meta.membership_entry_count {
        count_mismatch(context, ValidationObject::MembershipDictionary)?;
    }

    let mut previous_hash = None;
    let hash_result = tree::walk::<HashCodec, S, _>(
        context,
        context.meta.membership_hash_root,
        ValidationObject::MembershipReverseIndex,
        |context, page, cell| validate_hash(context, page, cell, &mut previous_hash),
    )?;
    if hash_result.records != context.meta.membership_entry_count {
        count_mismatch(context, ValidationObject::MembershipReverseIndex)?;
    }

    let used = bitmap::validate(
        context,
        context.meta.membership_used_root,
        context.meta.membership_id_limit,
        Kind::Membership,
    )?;
    finish(context, id_result.records, used, maximum_id)
}

fn validate_record<S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    cell: P,
    feeds: &mut WordCache,
    maximum_id: &mut u32,
) -> Result<()> {
    let Some(record) = decode_record(cell) else {
        return Ok(());
    };
    *maximum_id = (*maximum_id).max(record.id);
    if u64::from(record.id) >= context.meta.membership_id_limit {
        bitmap_finding(context, Some(page_number))?;
    }
    if record.refcount == 0 {
        refcount_finding(context, Some(page_number))?;
    }
    if !matches!(
        context.define_membership(record.id, record.refcount, record.word_count, record.digest,)?,
        InsertResult::Inserted
    ) {
        bitmap_finding(context, Some(page_number))?;
    }
    validate_record_bitmap(context, page_number, record, feeds)
}

fn validate_record_bitmap<S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    record: Record<P>,
    feeds: &mut WordCache,
) -> Result<()> {
    let mut scan = BitmapScan::new(feeds);
    let complete = match record.storage {
        Storage::Inline(bytes) => {
            scan.consume(context, bytes)?;
            true
        }
        Storage::Blob(root) => blob::scan_membership(
            context,
            root,
            u64::from(record.word_count) * 8,
            |context, bytes| scan.consume(context, bytes),
        )?,
    };
    let length_matches = scan.words == record.word_count;
    report_bitmap_shape(context, page_number, &scan, length_matches)?;
    report_active_feeds(context, page_number, scan.active_invalid)?;
    let digest = scan.finish_digest();
    report_digest(
        context,
        page_number,
        complete,
        length_matches,
        digest,
        record.digest,
    )
}

fn report_bitmap_shape<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    scan: &BitmapScan<'_>,
    length_matches: bool,
) -> Result<()> {
    if !length_matches || scan.last_word == 0 {
        bitmap_finding(context, Some(page_number))?;
    }
    Ok(())
}

fn report_active_feeds<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    invalid: bool,
) -> Result<()> {
    if invalid {
        context.emit(
            ValidationReason::MembershipActiveFeedInvalid,
            ValidationObject::MembershipDictionary,
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(())
}

fn report_digest<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    complete: bool,
    length_matches: bool,
    actual: [u8; 32],
    expected: [u8; 32],
) -> Result<()> {
    if complete && length_matches && actual != expected {
        context.emit(
            ValidationReason::MembershipHashInvalid,
            ValidationObject::MembershipDictionary,
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(())
}

struct BitmapScan<'a> {
    feeds: &'a mut WordCache,
    hasher: Sha256,
    words: u32,
    last_word: u64,
    active_invalid: bool,
    feed_reader_failed: bool,
}

impl<'a> BitmapScan<'a> {
    fn new(feeds: &'a mut WordCache) -> Self {
        Self {
            feeds,
            hasher: Sha256::new(),
            words: 0,
            last_word: 0,
            active_invalid: false,
            feed_reader_failed: false,
        }
    }

    fn consume<S: ValidationSink, P: ByteSource>(
        &mut self,
        context: &mut Context<'_, S>,
        bytes: P,
    ) -> Result<()> {
        context.checkpoint()?;
        if bytes.len() % 8 != 0 {
            return Err(Error::Corrupt(
                "validated membership chunk is not word aligned",
            ));
        }
        for offset in (0..bytes.len()).step_by(8) {
            let value = crate::contract::u64_le(bytes, offset);
            self.hasher.update(value.to_le_bytes());
            self.check_active(context, value);
            self.last_word = value;
            self.words = self.words.checked_add(1).ok_or(Error::ArithmeticOverflow(
                "validation membership word count",
            ))?;
        }
        Ok(())
    }

    fn check_active<S: ValidationSink>(&mut self, context: &Context<'_, S>, value: u64) {
        if self.feed_reader_failed {
            return;
        }
        match self.feeds.word(context, self.words) {
            Ok(active) => self.active_invalid |= value & !active != 0,
            Err(_) => {
                self.active_invalid = true;
                self.feed_reader_failed = true;
            }
        }
    }

    fn finish_digest(self) -> [u8; 32] {
        self.hasher.finalize().into()
    }
}

fn validate_hash<S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    cell: P,
    previous: &mut Option<HashKey>,
) -> Result<()> {
    let Some(key) = decode_hash(cell) else {
        return Ok(());
    };
    if u64::from(key.id) >= context.meta.membership_id_limit
        || !context.mark_membership_reverse(key.id, key.word_count, key.digest)?
    {
        reverse_finding(context, Some(page_number))?;
    }
    if previous
        .is_some_and(|prior| prior.digest == key.digest && prior.word_count == key.word_count)
    {
        compare_collision(
            context,
            page_number,
            previous.unwrap().id,
            key.id,
            key.word_count,
        )?;
    }
    *previous = Some(key);
    Ok(())
}

fn compare_collision<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    left: u32,
    right: u32,
    word_count: u32,
) -> Result<()> {
    match equal_memberships(context, left, right, word_count) {
        Ok(true) => context.emit(
            ValidationReason::MembershipHashInvalid,
            ValidationObject::MembershipReverseIndex,
            Some(page_number),
            None,
            None,
        ),
        Ok(false) => Ok(()),
        Err(_) => reverse_finding(context, Some(page_number)),
    }
}

fn equal_memberships<S: ValidationSink>(
    context: &Context<'_, S>,
    left: u32,
    right: u32,
    word_count: u32,
) -> Result<bool> {
    let left = open_membership(context, left)?;
    let right = open_membership(context, right)?;
    if left.word_count()? != word_count || right.word_count()? != word_count {
        return Ok(false);
    }
    compare_words(context, &left, &right, word_count)
}

fn open_membership<'a, S: ValidationSink>(
    context: &'a Context<'_, S>,
    id: u32,
) -> Result<membership_view::MembershipView<'a>> {
    membership_view::by_id(context.mapping, &context.meta, id, None)
}

fn compare_words<S: ValidationSink>(
    context: &Context<'_, S>,
    left: &membership_view::MembershipView<'_>,
    right: &membership_view::MembershipView<'_>,
    word_count: u32,
) -> Result<bool> {
    let mut left_words = [0; COMPARE_WORDS];
    let mut right_words = [0; COMPARE_WORDS];
    let mut start = 0u32;
    while start < word_count {
        context.checkpoint()?;
        let count = usize::try_from((word_count - start).min(COMPARE_WORDS as u32))
            .map_err(|_| Error::ArithmeticOverflow("validation membership comparison"))?;
        left.read_words(start, &mut left_words[..count])?;
        right.read_words(start, &mut right_words[..count])?;
        if left_words[..count] != right_words[..count] {
            return Ok(false);
        }
        start += count as u32;
    }
    Ok(true)
}

fn finish<S: ValidationSink>(
    context: &mut Context<'_, S>,
    dictionary_records: u64,
    used_bits: u64,
    maximum_id: u32,
) -> Result<()> {
    let defined = validate_slots(context)?;
    let expected_limit = expected_id_limit(maximum_id);
    if !totals_match(
        context,
        defined,
        dictionary_records,
        used_bits,
        expected_limit,
    ) {
        bitmap_finding(context, None)?;
    }
    Ok(())
}

fn validate_slots<S: ValidationSink>(context: &mut Context<'_, S>) -> Result<u64> {
    let mut defined = 0;
    let mut used = WordCache::new(
        context.meta.membership_used_root,
        context.meta.membership_id_limit,
        Kind::Membership,
    );
    for index in 0..context.membership_slots()? {
        context.checkpoint()?;
        let Some(slot) = context.membership_slot(index)? else {
            continue;
        };
        if slot.defined {
            defined += 1;
        }
        validate_slot(context, slot, &mut used)?;
    }
    Ok(defined)
}

fn expected_id_limit(maximum_id: u32) -> u64 {
    if maximum_id == 0 {
        1
    } else {
        u64::from(maximum_id) + 1
    }
}

fn totals_match<S: ValidationSink>(
    context: &Context<'_, S>,
    defined: u64,
    dictionary_records: u64,
    used_bits: u64,
    expected_limit: u64,
) -> bool {
    defined == dictionary_records
        && defined == context.meta.membership_entry_count
        && used_bits == defined
        && context.meta.membership_id_limit == expected_limit
}

fn validate_slot<S: ValidationSink>(
    context: &mut Context<'_, S>,
    slot: super::membership_table::Slot,
    used: &mut WordCache,
) -> Result<()> {
    validate_slot_refcount(context, slot)?;
    validate_slot_reverse(context, slot)?;
    validate_slot_used(context, slot, used)
}

fn validate_slot_refcount<S: ValidationSink>(
    context: &mut Context<'_, S>,
    slot: super::membership_table::Slot,
) -> Result<()> {
    if !slot.defined || slot.stored_refcount == 0 || slot.stored_refcount != slot.range_count {
        refcount_finding(context, None)?;
    }
    Ok(())
}

fn validate_slot_reverse<S: ValidationSink>(
    context: &mut Context<'_, S>,
    slot: super::membership_table::Slot,
) -> Result<()> {
    if slot.defined && !slot.reverse_seen {
        reverse_finding(context, None)?;
    }
    Ok(())
}

fn validate_slot_used<S: ValidationSink>(
    context: &mut Context<'_, S>,
    slot: super::membership_table::Slot,
    used: &mut WordCache,
) -> Result<()> {
    if slot.defined {
        let word = used.word(context, slot.id / 64).unwrap_or(0);
        if word & (1u64 << (slot.id % 64)) == 0 {
            bitmap_finding(context, None)?;
        }
    }
    Ok(())
}

fn count_mismatch<S: ValidationSink>(
    context: &mut Context<'_, S>,
    object: ValidationObject,
) -> Result<()> {
    context.emit(ValidationReason::RootCountInvalid, object, None, None, None)
}

fn bitmap_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::MembershipBitmapInvalid,
        ValidationObject::MembershipDictionary,
        page,
        None,
        None,
    )
}

fn reverse_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::MembershipReverseIndexInvalid,
        ValidationObject::MembershipReverseIndex,
        page,
        None,
        None,
    )
}

fn refcount_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::MembershipRefcountInvalid,
        ValidationObject::MembershipDictionary,
        page,
        None,
        None,
    )
}

File: v4/go/internal/validation/membership.go
package validation

// Membership validation (Rust validation/membership.rs): the ID and
// hash tree walks over the membership dictionary with the record
// bitmap scan (word content, active-feed window, trailing word, and
// digest proofs), the reverse-index collision compare, the used-bitmap
// count, and the slot refcount/reverse/used finish. The reverse tables
// come from the bounded value-kind tables charged in the context.

import (
	"crypto/sha256"
	stdhash "hash"

	"github.com/firehol/iprange/v4/go/internal/bitmap"
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/reader"
	"github.com/firehol/iprange/v4/go/internal/tree"
)

// validateMembership runs the membership validators (Rust
// membership::validate): only membership and structured databases carry
// a dictionary; every other value kind is empty.
func validateMembership(ctx *context) error {
	if ctx.meta.ValueKind != format.ValueKindMembership && ctx.meta.ValueKind != format.ValueKindStructured {
		return nil
	}
	feeds := newBitmapWordCache(ctx.meta.FeedUsedRoot, ctx.meta.FeedIndexLimit, bitmap.KindFeed)
	maximumID := uint32(0)
	idResult, err := walkTree(ctx, ctx.meta.MembershipIDRoot, ObjectMembershipDictionary, membershipIDCodec(), func(ctx *context, pageNumber uint32, cell []byte) error {
		return validateMembershipRecord(ctx, pageNumber, cell, &feeds, &maximumID)
	})
	if err != nil {
		return err
	}
	if idResult.records != ctx.meta.MembershipEntryCount {
		if err := membershipCountMismatch(ctx, ObjectMembershipDictionary); err != nil {
			return err
		}
	}
	var previous format.MembershipHashKey
	hasPrevious := false
	catalog := reader.NewImmutable(ctx.mapping, ctx.meta)
	hashResult, err := walkTree(ctx, ctx.meta.MembershipHashRoot, ObjectMembershipReverseIndex, membershipHashCodec(), func(ctx *context, pageNumber uint32, cell []byte) error {
		return validateMembershipHash(ctx, pageNumber, cell, &previous, &hasPrevious, catalog)
	})
	if err != nil {
		return err
	}
	if hashResult.records != ctx.meta.MembershipEntryCount {
		if err := membershipCountMismatch(ctx, ObjectMembershipReverseIndex); err != nil {
			return err
		}
	}
	used, err := validateBitmap(ctx, ctx.meta.MembershipUsedRoot, ctx.meta.MembershipIDLimit, bitmap.KindMembership)
	if err != nil {
		return err
	}
	return finishMembership(ctx, idResult.records, used, maximumID)
}

// membershipIDCodec is the Codec of the ID tree (Rust IdCodec: fixed
// branch entries, variable records on the leaves, and the bitmap class
// for undecodable leaves).
func membershipIDCodec() treeCodec {
	return treeCodec{
		branchType:    byte(format.PageTypeMembershipIDBranch),
		leafType:      byte(format.PageTypeMembershipIDLeaf),
		aux:           0,
		branchLayout:  format.FixedLayout(format.MembershipIDBranchSize),
		leafLayout:    format.VariableLayout(format.MembershipIDRecordMin, format.MaxMembershipIDRecord),
		branchInvalid: ReasonTreeOrderInvalid,
		leafInvalid:   ReasonMembershipBitmapInvalid,
		branchKey: func(cell []byte) (tree.Key, bool) {
			firstID, _, err := format.DecodeMembershipIDBranchFields(cell)
			if err != nil {
				return tree.Key{}, false
			}
			return tree.Key{Lo: uint64(firstID)}, true
		},
		branchChild: func(cell []byte) (uint32, bool) {
			_, child, err := format.DecodeMembershipIDBranchFields(cell)
			if err != nil {
				return 0, false
			}
			return child, true
		},
		leafKey: func(cell []byte) (tree.Key, bool) {
			record, err := format.DecodeMembershipRecord(cell)
			if err != nil {
				return tree.Key{}, false
			}
			return tree.Key{Lo: uint64(record.ID)}, true
		},
	}
}

// membershipHashCodec is the Codec of the reverse-index tree (Rust
// HashCodec: fixed 40-byte hash leaves, fixed 44-byte branch entries,
// and the hash class for both levels).
func membershipHashCodec() treeCodec {
	return treeCodec{
		branchType:    byte(format.PageTypeMembershipHashBranch),
		leafType:      byte(format.PageTypeMembershipHashLeaf),
		aux:           0,
		branchLayout:  format.FixedLayout(format.MembershipHashBranchSize),
		leafLayout:    format.FixedLayout(format.MembershipHashKeySize),
		branchInvalid: ReasonMembershipHashInvalid,
		leafInvalid:   ReasonMembershipHashInvalid,
		branchKey: func(cell []byte) (tree.Key, bool) {
			key, _, err := format.DecodeMembershipHashBranchFields(cell)
			if err != nil {
				return tree.Key{}, false
			}
			return membershipHashTreeKey(key), true
		},
		branchChild: func(cell []byte) (uint32, bool) {
			_, child, err := format.DecodeMembershipHashBranchFields(cell)
			if err != nil {
				return 0, false
			}
			return child, true
		},
		leafKey: func(cell []byte) (tree.Key, bool) {
			key, err := format.DecodeMembershipHashKey(cell)
			if err != nil {
				return tree.Key{}, false
			}
			return membershipHashTreeKey(key), true
		},
	}
}

// membershipHashTreeKey builds the ordered key of one hash record (Rust
// HashKey Ord: digest bytes, then word count, then id; the little-endian
// counts keep that byte order).
func membershipHashTreeKey(key format.MembershipHashKey) tree.Key {
	var raw [40]byte
	copy(raw[:32], key.Digest[:])
	format.PutU32(raw[32:36], key.WordCount)
	format.PutU32(raw[36:40], key.ID)
	return tree.RawKey(raw[:])
}

// validateMembershipRecord checks one ID-tree leaf record (Rust
// validate_record): the id-limit window, the nonzero refcount, the
// single define, and the record bitmap scan.
func validateMembershipRecord(ctx *context, pageNumber uint32, cell []byte, feeds *bitmapWordCache, maximumID *uint32) error {
	record, err := format.DecodeMembershipRecord(cell)
	if err != nil {
		// The codec leaf key already reported the undecodable class.
		return nil
	}
	if record.ID > *maximumID {
		*maximumID = record.ID
	}
	if uint64(record.ID) >= ctx.meta.MembershipIDLimit {
		if err := membershipBitmapFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	if record.Refcount == 0 {
		if err := membershipRefcountFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	result, err := ctx.defineMembership(record.ID, record.Refcount, record.WordCount, record.Digest)
	if err != nil {
		return err
	}
	if result != InsertInserted {
		if err := membershipBitmapFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	return validateMembershipRecordBitmap(ctx, pageNumber, cell, record, feeds)
}

// validateMembershipRecordBitmap scans the record bitmap words (Rust
// validate_record_bitmap: inline records consume their proved inline
// span, blob records scan their blob tree) and reports the shape,
// active-feed, and digest classes.
func validateMembershipRecordBitmap(ctx *context, pageNumber uint32, cell []byte, record format.MembershipRecord, feeds *bitmapWordCache) error {
	scan := newMembershipBitmapScan(feeds)
	var complete bool
	switch record.Storage {
	case format.MembershipStorageInline:
		bytes := cell[format.MembershipIDRecordMin : format.MembershipIDRecordMin+int(record.WordCount)*8]
		if err := scan.consume(ctx, bytes); err != nil {
			return err
		}
		complete = true
	case format.MembershipStorageBlob:
		var err error
		complete, err = scanMembership(ctx, record.BlobRoot, uint64(record.WordCount)*8, func(ctx *context, bytes []byte) error {
			return scan.consume(ctx, bytes)
		})
		if err != nil {
			return err
		}
	}
	lengthMatches := scan.words == record.WordCount
	if !lengthMatches || scan.lastWord == 0 {
		if err := membershipBitmapFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	if scan.activeInvalid {
		if err := ctx.emit(ReasonMembershipActiveFeedInvalid, ObjectMembershipDictionary, &pageNumber, nil, nil); err != nil {
			return err
		}
	}
	if complete && lengthMatches && scan.finishDigest() != record.Digest {
		if err := membershipDigestFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	return nil
}

// membershipBitmapScan accumulates the word stream of one record bitmap
// (Rust BitmapScan): the digest over the words, the word count, the
// last word, and the active-feed window check through the feed cache.
type membershipBitmapScan struct {
	feeds            *bitmapWordCache
	hasher           stdhash.Hash
	words            uint32
	lastWord         uint64
	activeInvalid    bool
	feedReaderFailed bool
}

func newMembershipBitmapScan(feeds *bitmapWordCache) *membershipBitmapScan {
	return &membershipBitmapScan{feeds: feeds, hasher: sha256.New()}
}

// consume feeds one byte slice of the bitmap stream (Rust
// BitmapScan::consume: word aligned, checkpointed, and hashed in the
// wire byte order).
func (s *membershipBitmapScan) consume(ctx *context, bytes []byte) error {
	if err := ctx.checkpoint(); err != nil {
		return err
	}
	if len(bytes)%8 != 0 {
		return &format.Error{Code: format.CodeFormatInvalid, Detail: "validated membership chunk is not word aligned"}
	}
	for offset := 0; offset < len(bytes); offset += 8 {
		value := format.U64(bytes[offset : offset+8])
		s.hasher.Write(bytes[offset : offset+8])
		if !s.feedReaderFailed {
			s.checkActive(ctx, value)
		}
		s.lastWord = value
		if s.words == ^uint32(0) {
			return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation membership word count"}
		}
		s.words++
	}
	return nil
}

// checkActive folds one word against the feed used bitmap (Rust
// BitmapScan::check_active; a feed-read defect poisons the rest of the
// scan).
func (s *membershipBitmapScan) checkActive(ctx *context, value uint64) {
	active, err := s.feeds.word(ctx, s.words)
	if err != nil {
		s.activeInvalid = true
		s.feedReaderFailed = true
		return
	}
	s.activeInvalid = s.activeInvalid || value&^active != 0
}

// finishDigest returns the sha256 over the consumed words (Rust
// BitmapScan::finish_digest).
func (s *membershipBitmapScan) finishDigest() [32]byte {
	var digest [32]byte
	copy(digest[:], s.hasher.Sum(nil))
	return digest
}

// validateMembershipHash checks one reverse-index leaf record (Rust
// validate_hash): the id-limit window, the reverse mark, and the
// adjacent same-digest collision compare.
func validateMembershipHash(ctx *context, pageNumber uint32, cell []byte, previous *format.MembershipHashKey, hasPrevious *bool, catalog *reader.ImmutableReader) error {
	key, err := format.DecodeMembershipHashKey(cell)
	if err != nil {
		// The codec leaf key already reported the undecodable class.
		return nil
	}
	marked, err := ctx.markMembershipReverse(key.ID, key.WordCount, key.Digest)
	if err != nil {
		return err
	}
	if uint64(key.ID) >= ctx.meta.MembershipIDLimit || !marked {
		if err := membershipReverseFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	if *hasPrevious && previous.Digest == key.Digest && previous.WordCount == key.WordCount {
		if err := compareMembershipCollision(ctx, pageNumber, previous.ID, key.ID, key.WordCount, catalog); err != nil {
			return err
		}
	}
	*previous = key
	*hasPrevious = true
	return nil
}

// compareMembershipCollision compares the two adjacent records of one
// digest (Rust compare_collision: equal bitmaps are the hash class,
// unequal are clean, and any read defect is the reverse class).
func compareMembershipCollision(ctx *context, pageNumber uint32, left, right, wordCount uint32, catalog *reader.ImmutableReader) error {
	equal, err := equalMemberships(ctx, catalog, left, right, wordCount)
	if err != nil {
		return membershipReverseFinding(ctx, &pageNumber)
	}
	if equal {
		return membershipHashFinding(ctx, &pageNumber)
	}
	return nil
}

// equalMemberships proves two records identical over their bitmap words
// (Rust equal_memberships).
func equalMemberships(ctx *context, catalog *reader.ImmutableReader, left, right, wordCount uint32) (bool, error) {
	leftView, err := catalog.LookupMembershipID(left)
	if err != nil {
		return false, err
	}
	rightView, err := catalog.LookupMembershipID(right)
	if err != nil {
		return false, err
	}
	if leftView.WordCount() != wordCount || rightView.WordCount() != wordCount {
		return false, nil
	}
	return compareMembershipWords(ctx, leftView, rightView, wordCount)
}

// membershipCompareWords is the bounded chunk of one word comparison
// (Rust COMPARE_WORDS).
const membershipCompareWords = 64

// compareMembershipWords compares the two bitmaps chunk by chunk (Rust
// compare_words).
func compareMembershipWords(ctx *context, left, right reader.MembershipView, wordCount uint32) (bool, error) {
	var leftWords [membershipCompareWords]uint64
	var rightWords [membershipCompareWords]uint64
	start := uint32(0)
	for start < wordCount {
		if err := ctx.checkpoint(); err != nil {
			return false, err
		}
		count := wordCount - start
		if count > membershipCompareWords {
			count = membershipCompareWords
		}
		if _, err := left.ReadWords(start, leftWords[:count]); err != nil {
			return false, err
		}
		if _, err := right.ReadWords(start, rightWords[:count]); err != nil {
			return false, err
		}
		for i := 0; i < int(count); i++ {
			if leftWords[i] != rightWords[i] {
				return false, nil
			}
		}
		start += count
	}
	return true, nil
}

// finishMembership proves the dictionary totals (Rust finish: the slot
// defined count, the used-bitmap bits, and the exact id limit against
// the maximum observed id).
func finishMembership(ctx *context, dictionaryRecords, usedBits uint64, maximumID uint32) error {
	defined, err := validateMembershipSlots(ctx)
	if err != nil {
		return err
	}
	expectedLimit := uint64(1)
	if maximumID != 0 {
		expectedLimit = uint64(maximumID) + 1
	}
	if defined != dictionaryRecords || defined != ctx.meta.MembershipEntryCount ||
		usedBits != defined || ctx.meta.MembershipIDLimit != expectedLimit {
		return membershipBitmapFinding(ctx, nil)
	}
	return nil
}

// validateMembershipSlots walks every table entry (Rust validate_slots).
func validateMembershipSlots(ctx *context) (uint64, error) {
	defined := uint64(0)
	used := newBitmapWordCache(ctx.meta.MembershipUsedRoot, ctx.meta.MembershipIDLimit, bitmap.KindMembership)
	slots, err := ctx.membershipSlots()
	if err != nil {
		return 0, err
	}
	for index := 0; index < slots; index++ {
		if err := ctx.checkpoint(); err != nil {
			return 0, err
		}
		slot, ok, err := ctx.membershipSlot(index)
		if err != nil {
			return 0, err
		}
		if !ok {
			continue
		}
		if slot.Defined {
			defined++
		}
		if err := validateMembershipSlot(ctx, slot, &used); err != nil {
			return 0, err
		}
	}
	return defined, nil
}

// validateMembershipSlot checks one occupied table entry (Rust
// validate_slot: refcount, reverse mark, and used bit).
func validateMembershipSlot(ctx *context, slot Slot, used *bitmapWordCache) error {
	if err := validateMembershipSlotRefcount(ctx, slot); err != nil {
		return err
	}
	if err := validateMembershipSlotReverse(ctx, slot); err != nil {
		return err
	}
	return validateMembershipSlotUsed(ctx, slot, used)
}

func validateMembershipSlotRefcount(ctx *context, slot Slot) error {
	// The Rust arm fires for every occupied slot that is not defined by
	// the dictionary too: a range-counted id without a dictionary
	// record is as invalid as a stored/counted mismatch.
	if !slot.Defined || slot.StoredRefcnt == 0 || slot.StoredRefcnt != slot.RangeCount {
		return membershipRefcountFinding(ctx, nil)
	}
	return nil
}

func validateMembershipSlotReverse(ctx *context, slot Slot) error {
	if slot.Defined && !slot.ReverseSeen {
		return membershipReverseFinding(ctx, nil)
	}
	return nil
}

func validateMembershipSlotUsed(ctx *context, slot Slot, used *bitmapWordCache) error {
	if slot.Defined {
		word, err := used.word(ctx, slot.ID/64)
		if err != nil {
			word = 0
		}
		if word&(uint64(1)<<(slot.ID%64)) == 0 {
			return membershipBitmapFinding(ctx, nil)
		}
	}
	return nil
}

func membershipCountMismatch(ctx *context, object ValidationObject) error {
	return ctx.emit(ReasonRootCountInvalid, object, nil, nil, nil)
}

func membershipBitmapFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonMembershipBitmapInvalid, ObjectMembershipDictionary, page, nil, nil)
}

func membershipReverseFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonMembershipReverseIndexInvalid, ObjectMembershipReverseIndex, page, nil, nil)
}

func membershipRefcountFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonMembershipRefcountInvalid, ObjectMembershipDictionary, page, nil, nil)
}

// membershipHashFinding streams the hash class of the reverse-index
// tree object (Rust compare_collision; the hash record walk itself
// reports the same class on the reverse-index object).
func membershipHashFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonMembershipHashInvalid, ObjectMembershipReverseIndex, page, nil, nil)
}

// membershipDigestFinding streams the hash class of a dictionary record
// whose stored digest disagrees with its bitmap (Rust report_digest).
func membershipDigestFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonMembershipHashInvalid, ObjectMembershipDictionary, page, nil, nil)
}
